### ISSUE TITLE
Videos do not load in Safari

### ISSUE THREAD

**Issue by angelikatyborska** [unpoly/unpoly#432]
It's me again, the author of https://github.com/unpoly/unpoly/issues/175 with more Safari-related problems 😢.

I am aware that this is most likely Safari's bug and there's little we can do about it, but I thought i will report it here anyway in the hopes of brainstorming a better hack to work around it and warning others about it.

### Bug description

Videos do not load at all in Safari on pages visited with Unpoly links.

### Reproduction project

- [GitHub repo unpoly_video_test](https://github.com/angelikatyborska/unpoly_video_test)
- [Glitch project](https://glitch.com/edit/#!/silky-wistful-coast) (same content as the GitHub repo minus the README)

### Steps to reproduce the behavior:

This behavior is recorded as a gif and included in the reproduction project GitHub repo's README.

1. Go to reproduction project linked above **in macOS Safari**
2. Start the preview, `index.html` loads with an autoplaying video
3. Click link "page 1" to reload the same page via Unpoly
4. The video disappears (it should still be there, autoplaying)
5. Click link "page 2" to go to a similar page with an autoplaying video with controls
6. There is no video (it should be there, autoplaying with controls)
7. Edit the path of the preview to `index2.html` to reload the second page
8. There is a video, as it should

### Expected behavior

Videos should always load when navigating with Unpoly, respecting the desired autoplay/mute/control options, in all modern browsers.

### Browser version

 - macOS 12.4
 - Safari Version 15.5 (17613.2.7.1.8)

and also

- iOS 15.5 on an iPad
- Safari

### Additional context

This bug doesn't appear for me in other browsers that I tested (macOS Firefox, macOS Chrome, Android Firefox, Android Chrome).

I found a dirty workaround for autoplaying videos without controls, but it doesn't work for videos with controls - the controls never load.

```js
up.compiler('video', async function (video) {
  try {
    await video.play()
  } catch (err) {
    // do nothing
  }
})
```

This would be good enough for me, but unfortunately I do have a not-autoplaying video with controls on my current project that uses Unpoly and I'm looking for any ideas on how to approach this problem. Currently I'm thinking that the only solution to this problem might be to somehow detect with [HTMLMediaElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement)'s API that the video didn't load at all, and replace it with an image and a link to the video then.

**Comment by triskweline** [comment]
I haven't investigated this deeply, but I think we sometimes have a compiler that re-creates the `<video>` element at runtime for Safari.

In the HTML the `<video>` element is within a `<noscript>` tag:

```html
<div class="video-player">
  <noscript>
    <video controls autoplay>
      <source type="..." src="...">
    </video>
  </noscript>
</div>
```

Then we create that `<video>` element from a compiler:

```js
up.compiler('.video-player', function(wrapper) {
  const noscript = wrapper.querySelector('noscript')
  const videoHTML = noscript.innerText.trim()
  const video = up.element.createFromHTML(videoHTML)
  wrapper.appendChild(video)
})
```

**Comment by adam12** [comment]
Still valid. [Remixed Glitch](https://glitch.com/edit/#!/spiky-suave-train) with a working video file.

**Comment by binaryfire** [comment]
Just wondering if there’s an update on this? Interested in trying Unpoly as an alternative to HTMX for a video-heavy site.

**Comment by adam12** [comment]
@binaryfire Remixed glitch still shows the issue, but I believe Henning posted a workaround above, so you should be OK.

**Comment by triskweline** [comment]
This looks like an issue in Safari, not specific to Unpoly:

- Related htmx issue: [Safari (web and iOS) doesn't re-create / copy the ShadowRoot when a web component is swapped](https://github.com/bigskysoftware/htmx/issues/764)
- Related Webkit issue: [Media element never populates its UA shadow if it was initially created in a document without browsing context](https://bugs.webkit.org/show_bug.cgi?id=222657)

Some ideas for addressing this (in addition to the [workaround above](https://github.com/unpoly/unpoly/issues/432#issuecomment-1263449773)):

- Maybe Unpoly could use a different way to parse HTML documents so the new elements belong to the current `document` (or browsing context). We're currently using `DOMParser` for that.
- Maybe it would help if Unpoly would explictly call `document.adoptNode()` on new elements in `up.ResponseDoc#finalizeElement()`.

> Interested in trying Unpoly as an alternative to HTMX for a video-heavy site.

It looks like htmx has the same issue.

My team built two high-traffic video sites ([Audi MediaTV](https://www.audi-mediacenter.com/en/videos), [Studyflix](https://studyflix.de/)) using Unpoly.

**Comment by triskweline** [comment]
> Maybe Unpoly could use a different way to parse HTML documents so the new elements belong to the current document (or browsing context). We're currently using DOMParser for that.

I actually got this working in [hk/parse-html-in-current-document](https://github.com/unpoly/unpoly/tree/hk/parse-html-in-current-document), by parsing HTML like this:

```js
let htmlElement = document.createElement('html')
htmlElement.innerHTML = html
```

Huge performance footgun here: This immediately instantiates all custom elements in the given `html`, even when we only end up inserting some of these instances. It may not be such a bad thing to parse HTML as an external document first, even if it means dealing with some warts like `<video>` on Safari and the existing handling we have for `<script>` and `<noscript>` elements.

**Comment by triskweline** [comment]
Fixed in 3.5.2.
